llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -407,6 +410,23 @@ Expected<GroupRef> GroupRef::create(MCCASBuilder &MB,
   return get(B->build());
 }
 
+template <typename RefTy>
+static Expected<SmallVector<RefTy, 0>> findRefs(MCCASReader &Reader,
+                                                ArrayRef<cas::ObjectRef> Refs) {
+  SmallVector<RefTy, 0> TopRefs;
+  for (auto ID : Refs) {
+    auto Node = Reader.getObjectProxy(ID);
+    if (!Node)
+      return Node.takeError();
+    if (auto TopRef = RefTy::Cast(*Node))
+      TopRefs.push_back(*TopRef);
+  }
+  if (TopRefs.size())
+    return TopRefs;
+  return createStringError(inconvertibleErrorCode(),
+                           "failed to find reference");
+}
```

Comment:
There is quite a bit duplication with `findRef` here. To avoid some of that, could you also rewrite `findRef` in terms of your new function?

```
template <typename RefTy>
static Expected<RefTy> findRef(MCCASReader &Reader,
                               ArrayRef<cas::ObjectRef> Refs) {
  auto Refs = findRefs<RefTy>(Reader, Refs);
  if (!Refs) return Refs.takeError();
  return Refs->front();
}
```